offchain/dispatcher/src/drivers/context.rs

```diff
@@ -62,98 +67,122 @@ impl Context {
         input: &Input,
         broker: &impl BrokerSend,
     ) -> Result<(), BrokerFacadeError> {
-        broker.enqueue_input(self.inputs_sent_count, input).await?;
+        let input_block = input.block_added.number.as_u64();
+        self.finish_epoch_if_needed(input_block, broker).await?;
```

Comment:
The other one reacts to a random block at the end of all the enqueues.
It is doing something conceptually different.

The deal with `finish_epoch_if_needed` inside the enqueue has to do with the consistency of the `Context`.
Imagine if we didn't have it inside the enqueue. Now let's look at the following order of events:
- enqueue(input from block 0)
- enqueue(input from block 1)
- enqueue(input from block 2)

Given that the epoch length is 2, and genesis  is 0, this would create 3 consecutive inputs without closing the epoch in between, which taints `last_input_epoch` and `last_finished_epoch`.

Thus, because this relates to the internal consistency of the `Context` module, it should belong to the `Context` module.
Before, this logic was divided between `context.rs` and `machine.rs`, which makes it difficult to reason with (and even test).